quota/quota.go

```diff
@@ -0,0 +1,315 @@
+		parentGroup: grp,
+	}
+
+	if err := subGrp.validate(); err != nil {
+		return nil, err
+	}
+
+	// also double check that the sub group name is not the same as that of the
+	// parent, this is fine in systemd world, but in snapd we want unique quota
+	// groups
+	if name == grp.Name {
+		return nil, fmt.Errorf("cannot use same name %q for sub group as parent group", name)
+	}
+
+	// save the details of this new sub-group in the parent group
+	grp.subGroups = append(grp.subGroups, subGrp)
+	grp.SubGroups = append(grp.SubGroups, name)
+
+	return subGrp, nil
+}
+
+func NewGroup(name string, memLimit quantity.Size) (*Group, error) {
+	grp := &Group{
+		Name:        name,
+		MemoryLimit: memLimit,
+	}
+
+	if err := grp.validate(); err != nil {
+		return nil, err
+	}
+
+	return grp, nil
+}
+
+// CompleteCrossReferences takes a set of deserialized groups and sets all
+// cross references amongst them using the unexported fields which are not
+// serialized.
+func CompleteCrossReferences(grps map[string]*Group) error {
+	// iterate over all groups, looking for sub-groups which need to be threaded
+	// together with their respective parent groups from the set
+	for _, grp := range grps {
+		// first thread the parent link
+		if grp.ParentGroup != "" {
+			parent, ok := grps[grp.ParentGroup]
+			if !ok {
+				return fmt.Errorf("internal error: missing group %q referenced as the parent of group %q", grp.ParentGroup, grp.Name)
+			}
+			grp.parentGroup = parent
+		}
+
+		// now thread any child links from this group to any children
+		if len(grp.SubGroups) != 0 {
+			grp.subGroups = make([]*Group, len(grp.SubGroups))
+			for i, subName := range grp.SubGroups {
+				sub, ok := grps[subName]
+				if !ok {
+					return fmt.Errorf("internal error: missing group %q referenced as the child of group %q", subName, grp.Name)
+				}
+				grp.subGroups[i] = sub
+			}
+		}
+	}
+
+	return nil
+}
+
+// subTree recursively returns all of the sub-groups of the group
+func (grp *Group) subTree() []*Group {
+	subTreeList := grp.subGroups
+	for _, sub := range grp.subGroups {
+		subTreeList = append(subTreeList, sub.subTree()...)
+	}
+
+	return subTreeList
+}
+
+func deleteGrpFromList(grps []*Group, grpToDelete *Group) []*Group {
+	grpCp := make([]*Group, 0, len(grps))
+	for _, grp := range grps {
+		if grp != grpToDelete {
+			grpCp = append(grpCp, grp)
+		}
+	}
+	return grpCp
+}
+
+type QuotaGroupSet struct {
+	grps map[*Group]bool
+}
+
+// AddAllNecessaryGroups adds all groups that are required for the specified
+// group to be effective to the set. This means all sub-groups of this group,
+// all parent groups of this group, and all sub-trees of any parent groups. This
+// set is the set of quota groups that must exist for this quota group to be
+// fully realized on a system, since all sub-branches of the full tree must
+// exist since this group may share some quota resources with the other
+// branches.
+func (s *QuotaGroupSet) AddAllNecessaryGroups(grp *Group) {
```

Comment:
whouldn't it work to instead get to the top parent and mark the subTree from there?

if I understand the usage of this in this PR it might get called to readd the same groups again and aganin? it should have shortcuts when it sets things to true to return/break if they were already true?

Reply:
Yes indeed just walking up to the root is a simpler implementation; given that we don't expect to have extremely large trees, I think it's fine to not worry about shortcutting like you describe and it's better to just have a simple implementation for now that does the right thing, even if it is not super efficient.

Reply:
This is done now